### PR TITLE
Remove dollar sign from label in variable completion response

### LINKED ISSUE #904: LS: Double dollar sign in completion response
- use phpactor as language server
- open any class with a function body e.g. `lib/Extension/Core/Console/PhpactorCommandLoader.php`
- inside the body write `$thi` and select `$this`.

Additional dollar sign appears then in response. 
This error does not occur when perfoming the same completion using omnifunc.

```
[Trace - 00:25:50] Sending notification 'textDocument/didChange'.
Params: {
    "textDocument": {
        "uri": "file:///home/user/.vim/bundle/phpactor/lib/Extension/Core/Console/PhpactorCommandLoader.php",
        "version": 6
    },
    "contentChanges": [
        {
            "text": "<?php\n\nnamespace Phpactor\\Extension\\Core\\Console;\n\nuse Symfony\\Component\\Console\\CommandLoader\\ContainerCommandLoader;\n\nclass PhpactorCommandLoader extends ContainerCommandLoader\n{\n    public function get($name)\n    {\n        $$this\n        $command = parent::get($name);\n        $command->setName($name);\n\n        return $command;\n    }\n}\n"
        }
    ]
}
``

### PR DESCRIPTION
**Pull request by przepompownia**
Closes https://github.com/phpactor/phpactor/issues/904

### CODE REVIEW

File: lib/Handler/CompletionHandler.php [hunk 1]
```diff
@@ -96,10 +96,21 @@ public function completion(TextDocumentItem $textDocument, Position $position):
...
         $capabilities->signatureHelpProvider = new SignatureHelpOptions(['(', ',']);
     }
 
+    private function formatLabel(Suggestion $suggestion): string
+    {
+        switch ($suggestion->type()) {
+            case Suggestion::TYPE_VARIABLE:
+                return mb_substr($suggestion->label(), 1);
+            default:
+                return $suggestion->label();
```
Comment (dantleech) [inline lib/Handler/CompletionHandler.php]: I would remove `default` and move the return to be after the switch statement (so the switch is a guard I guess). Could also possibly extract the "formatter" to a new class possibly -- if it makes it easier to test at this stage.

**Comment by dantleech** [comment]
Probably worth adding a test for this: https://github.com/phpactor/language-server-completion-extension/blob/master/tests/Unit/Handler/CompletionHandlerTest.php#L65

**Comment by przepompownia** [comment]
Probably in place of simple `SuggestionLabelFormatter` should be `CompletionItemFactory` because of more complex cases like function parameters completion. What do you think about it (before I would add some tests to `SuggestionLabelFormatter`)?

File: tests/Unit/Handler/SignatureHelpHandlerTest.php [hunk 2]
```diff
@@ -35,7 +35,7 @@ class SignatureHelpHandlerTest extends TestCase
      */
     private $workspace;
 
-    public function setUp()
+    public function setUp(): void
```
Comment (dantleech) [inline tests/Unit/Handler/SignatureHelpHandlerTest.php:38]: note: should be `protected` technically - but not your introduction :)

**Comment by dantleech** [comment]
> Probably in place of simple SuggestionLabelFormatter

I assume the goal is to separate the "formatters" - i.e. to have a method formatter, variable formatter etc? 

Not sure what a factory implementation would look like - I _might_ create a `Formatter` intertface and add a chain formatter which tried all formatters until one returns not null or similar. But might be over-kill. Maybe your idea is simpler.

Equally, if we can't see any great increase in complexity what we have now is fine.

**Comment by przepompownia** [comment]
Yes, we have consider it as done if can accept at the moment `switch` instead of splitting into separate formatters. 

I meant something different i.e. suspected that we can have to put somewhere any logic between e.g.  label and textEdit, but there is still no need to create such factory.

File: lib/Handler/CompletionHandler.php [hunk 3]
```diff
@@ -80,7 +88,7 @@ public function completion(TextDocumentItem $textDocument, Position $position):
         foreach ($suggestions as $suggestion) {
             /** @var Suggestion $suggestion */
             $completionList->items[] = new CompletionItem(
-                $suggestion->name(),
+                $this->suggestionLabelFormatter->format($suggestion),
```
Comment (dantleech) [inline lib/Handler/CompletionHandler.php]: hmm, why do we replace `name` with `label`? Should this be `name` formatter?

Comment (przepompownia) [reply lib/Handler/CompletionHandler.php]: Thanks for attention. It came from the first trying and was left untouched, even when moved to formatter ;)

Indeed, for example in `WorseClassMemberCompletor` a method name is mapped to suggestion name. I am surprised how then version using label works as expected.

**Comment by dantleech** [comment]
:boom:

File: lib/Util/SuggestionNameFormatter.php [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        $name = $suggestion->name();
+
+        switch ($suggestion->type()) {
+            case Suggestion::TYPE_VARIABLE:
+                $name = mb_substr($name, 1);
+                break;
+            case Suggestion::TYPE_FUNCTION:
+            case Suggestion::TYPE_METHOD:
+                $name = $name . '(';
```
Comment (dantleech) [inline lib/Util/SuggestionNameFormatter.php:19]: I think we should remove this also: `(` is used as to trigger signature help.

**Comment by przepompownia** [comment]
> I think we should remove this also: ( is used as to trigger signature help.

Probably it should be replaced by a Command that (after inserting) request the server to show signature help. For example, Intelephense's completion item has
```json
            "command": {
                "title": "Trigger Parameter Hints",
                "command": "editor.action.triggerParameterHints"
            },
```
https://microsoft.github.io/language-server-protocol/specification#command

File: lib/Util/SuggestionNameFormatter.php (hunk 4 above)
Comment (przepompownia) [reply lib/Util/SuggestionNameFormatter.php:19]: You are right. It is a hack although personally obtaining `(` is for me more practical than displaying each time signature help. I have defined a mapping to trigger the help:
```vim
" abstraction to reuse later by any client, e.g. phpactpr RPC
imap <buffer> <C-\>, <Plug>(ide-show-signature-help)
" in the case of using CoC
imap <Plug>(ide-show-signature-help) <C-o>:call CocActionAsync('showSignatureHelp')<CR>
```

**Comment by dantleech** [comment]
hmm, yeah that could be an option. Phpstorm will add `()` and put the cursor inside, which might also be an option with a text edit.

would also be quite trivial to add the command.

**Comment by dantleech** [comment]
I added it to the board: https://github.com/phpactor/phpactor/projects/1

**Comment by przepompownia** [comment]
Currently, I use https://github.com/jiangmiao/auto-pairs.git which makes its job except some small edge cases.

**Pull request merged**